docs/proposals/003-applications-outside-argocd-namespace.md

```diff
@@ -0,0 +1,418 @@
+  or components.
+
+### Non-Goals
+
+* Allow reconciliation from Argo CD `Application` resources that are created in
+  remote clusters. We believe this would be possible with some adaptions, but
+  for this initial proposal we only consider the control plane's cluster as a
+  source for `Application` resources.
+
+* Allow users to create `AppProject` resources on their own. Since `AppProject`
+  is used as central entity for enforcing governance and security, these should
+  stay in the control of the cluster administrators.
+
+* Replace or modify Argo CD internal RBAC model
+
+## Proposal
+
+We suggest to adapt the current mechanisms of reconciliation of `Application`
+resources to include resources within the control plane's cluster, but outside
+the control plane's namespace (e.g. `argocd`).
+
+We think the following changes would need to be performed so that this will
+become possible in a secure manner:
+
+* Adapt the `AppProject` spec to include a new field, e.g. `sourceNamespaces`.
+  The values of this new field will define which applications will be allowed
+  to associate to the `AppProject`. Consider the following example:
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: AppProject
+  metadata:
+    name: some-project
+    namespace: argocd
+  spec:
+    sourceNamespaces:
+    - foo-ns
+    - bar-ns
+  ```
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: Application
+  metadata:
+    name: some-app
+    namespace: bar-ns
+  spec:
+    project: some-project
+  ```
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: Application
+  metadata:
+    name: other-app
+    namespace: other-ns
+  spec:
+    project: some-project
+  ```
+
+  would allow `Application` resources that are created in either namespace
+  `foo-ns` or `bar-ns` to specify `some-project` in their `.spec.project`
+  field to associate themselves to the `AppProject` named `some-project`.
+  In the above example, the Application `some-app` would be allowed to associate
+  to the AppProject `some-project`, but the Application `other-app` would be
+  invalid.
+
+  This method would allow to delegate certain namespaces where users have
+  Kubernetes RBAC access to create `Application` resources.
+
+* `Applications` created in the control-plane's namespace (e.g. `argocd`) are
+  allowed to associate with any project when created declaratively, as they
+  are considered created by a super-user. So the following example would be
+  allowed to associate itself to project `some-project`, even with the
+  `argocd` namespace not being in the list of allowed `sourceNamespaces`:
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: Application
+  metadata:
+    name: superuser-app
+    namespace: argocd
+  spec:
+    project: some-project
+  ```
+
+* `Applications` created imperatively (e.g. through Argo CD API via UI or
+  CLI) will keep being created in the control plane namespace, and Argo CD
```

Comment:
So what if users define apps in new ns, in this case they can create with `kubectl -f file.yaml` and they are not able to create the same manifest with `argocd app create -f file.yaml` right? wouldn't it be confusing?

Reply:
I think the corresponding command would then be `argocd app create -n <namespace> -f file.yaml` (or specfiying the target namespace in the YAML). The API would then show this application as `<namespace>/<appname>` instead of just `<appname>`.